maintenance: skip bootout/bootstrap when plist is registered

derrickstolee opened the pull request:
This is by recommendation of @dscho. This should hopefully reduce concurrency problems in the Scalar functional tests, but it might not be enough for the very first few concurrent clones.

derrickstolee committed:
maintenance: skip bootout/bootstrap when plist is registered

On macOS, we use launchctl to manage the background maintenance
schedule. This uses a set of .plist files to describe the schedule, but
these files are also registered with 'launchctl bootstrap'. If multiple
'git maintenance start' commands run concurrently, then they can collide
replacing these schedule files and registering them with launchctl.

To avoid extra launchctl commands, do a check for the .plist files on
disk and check if they are registered using 'launchctl list <name>'.
This command will return with exit code 0 if it exists, or exit code 113
if it does not.

We can test this behavior using the GIT_TEST_MAINT_SCHEDULER environment
variable.

Signed-off-by: Derrick Stolee <dstolee@microsoft.com>
Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho commented on builtin/gc.c:
```diff
@@ -1615,15 +1615,46 @@ static int launchctl_remove_plists(const char *cmd)
+	if (!stat(filename, &st) && st.st_size) {
+		/* Check if it is registered */
+		if (launchctl_list_contains_plist(name, cmd))
+			goto cleanup;
```
Should we first verify that the contents are as expected? Otherwise, an updated `git maintenance` (that wants to write a slightly changed `.plist`) might not work as expected if the task had been registered previously?

dscho replied:
I implemented this.

dscho commented on t/t7900-maintenance.sh:
```diff
@@ -578,6 +578,23 @@ test_expect_success 'start and stop macOS maintenance' '
+	GIT_TEST_MAINT_SCHEDULER=launchctl:./print-args git maintenance start &&
+
+	# do it again on a fresh args file
+	rm -f args &&
```
Maybe we should also include `"$HOME/Library/LaunchAgents"` in that `rm -f` call?

dscho replied:
My mistake, that is just a directory, and while we list its contents into `actual`, we do not really do anything with it, it's just for debugging, I assume.

dscho commented on builtin/gc.c:
```diff
@@ -1615,15 +1615,46 @@ static int launchctl_remove_plists(const char *cmd)
+
 	if (safe_create_leading_directories(filename))
 		die(_("failed to create directories for '%s'"), filename);
 	plist = xfopen(filename, "w");
```
I would feel more comfortable if we called `hold_lock_file_for_update_timeout(&lock, filename, timeout_ms)` here.

And once the lock was acquired, not only write the file (i.e. do the very same stuff we do until that `fclose(plist)`, which would now be a `close_lock_file_gently(&lock)`), but _then_ verify that the plist is registered, and _then_ verify that the file contents are the same, and only if all those conditions hold true, `rollback_lock_file(&lock)` and `goto cleanup`, otherwise `commit_lock_file(&lock)` and do the `launchctl` dance.

What do you think?

dscho replied:
Since you indicated privately that you'd be fine with me implementing this, I went ahead and force-pushed.

dscho committed:
maintenance: create `launchctl` configuration using a lock file

When two `git maintenance` processes try to write the `.plist` file, we
need to help them with serializing their efforts.

The 150ms time-out value was determined from thin air.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho reviewed (approved):
Let's try this.

dscho commented:
I am upstreaming this via https://github.com/gitgitgadget/git/pull/1024.
